### ISSUE TITLE
Planning: Produce breakdown of tasks for requests to register Organisations

### ISSUE THREAD

**Issue by pjmonks** [MauroDataMapper/mdm-explorer#430]
# Overview

The workflow for how new researcher users register and start organisations must be determined. This could also be considered a "first time" registration setup process.

This issue will be used to gather initial notes/requirements and to discuss how to proceed so that concrete implementation tasks can be defined. This is a "living" document and is subject to change.

# Breakdown

- [x] Identify requirements for what is needed to have a researcher join an organisation
- [x] Plan organisation joining/registration workflow with OUH

# Requirements

Based on the description taken from the epic #421 and discussions with OUH, the following requirements can be implied:

1. An anonymous user must be able to submit an organisation registration request to the SDE.
2. Organisation registration requests will be submitted only when an existing organisation/department does not exist for a new user. Otherwise, the anonymous user will follow the workflow from #429.
3. An organisation registration request _must_ allow no user ID to be recorded against it initially. _However_, once the first (approver) user account has been created in the SDE, the request _must_ be updated to map the user to the request.
4. An organisation registration request cannot be marked as complete until the user ID of the new user has been assigned to it.

# Organisation Registration

## When this happens

It is necessary to submit an _organisation registration request_ when:

- A new user attempts to submit a _user registration request_
- However, they cannot find a pre-selectable organisation and/or department for such a request
- Meaning they instead submit an _organisation registration request_ - with their personal details to act as the first organisation member (the "approver" by default).

## Sequence of Actions

Below is the intended workflow of actions to carry out as part of the organisation registration request:

```mermaid
flowchart  
    subgraph explorer
        anonUser[Anonymous user]
        firstTimeSignIn[First time sign-in]
        reviewRequestStatus[Review request status]
        userReadyForSde[Ready for SDE]
    end
    subgraph sde-admin
        pendingRegistration[Pending registration request]
        assignedRegistration[Assigned registration request]
        initialUserChecks[Initial user checks]
        rejectUser[Reject and notify user]
        mapUserToRequest[Map user to request]
        verifyOrg[Verify organisation]
        rejectOrg[Reject organisation]
        createOrgDept[Create organisation/department]
        createSdeUser[Create Mauro/SDE user records - approver]
        newUserReady[New user ready - approver]
        mapUserToOrg[Map user to organisation]
        mapUserToDept[Map user to department]

        pendingRegistration -- SDE administrator pickup --> assignedRegistration
        assignedRegistration --> initialUserChecks
        initialUserChecks -- Verified --> createSdeUser 
        initialUserChecks -- Rejected --> rejectUser
        createSdeUser --> mapUserToRequest
        mapUserToRequest --> newUserReady
        newUserReady --> verifyOrg
        verifyOrg -- Rejected --> rejectOrg
        verifyOrg -- Verified --> createOrgDept
        createOrgDept --> mapUserToOrg
        mapUserToOrg --> mapUserToDept
    end
    subgraph azure
        addAzureAccount[Add guest account to tenant]
        authenticateUser[Authenticate user]
    end

    anonUser -- Submit registration form --> pendingRegistration
    createSdeUser --> addAzureAccount
    addAzureAccount --> newUserReady
    newUserReady -- Notify user --> firstTimeSignIn
    firstTimeSignIn -- Authenticate --> authenticateUser
    authenticateUser --> reviewRequestStatus
    reviewRequestStatus --> userReadyForSde
    mapUserToDept --> userReadyForSde
```

## Request State Workflow

Below are some changes to the existing SDE request state workflow that will need to be achieved to support the additional request types for user and organisation registration:

```mermaid
stateDiagram-v2
    draft : Draft request
    awaitUserCheck : Await user checks
    initialUserCheck : Initial user check
    createUserAccount : Create user account
    awaitPickup : Await pickup
    otherSdeStates : Other SDE states...

    draft --> awaitUserCheck : Submit (Researcher)

    awaitUserCheck --> initialUserCheck : Assign (SDE Administrator)
    initialUserCheck --> draft : Rejected (SDE Administrator)
    initialUserCheck --> createUserAccount : Approved (SDE Administrator)
    createUserAccount --> awaitPickup : User ready (SDE Administrator)
    createUserAccount --> initialUserCheck : Pending (SDE Administrator)
    createUserAccount --> awaitUserCheck : Pending (SDE Administrator)

    awaitPickup --> otherSdeStates : Continue... (SDE Administrator)

```

The key differences are what happens between "Draft Request" and "Await Approval" or "Await Pickup":

- For this kind of request, the status transitions straight from "Draft Request" to "Await User Checks".
- Only SDE administrators can manage requests in the "Await User Checks" state.
- An SDE administrator must be assigned to the request, moving to the "Initial User Checks" state.
- In this state, the SDE administrator must make some initial background checks on the user to ensure they should have a user account. (These are processes to be determined by OUH).
- All being well, they move the request to the "Create User Account" state. Here, they will perform the technical actions to setup their user account. Once created, the request can finally be mapped to this new user account.
- Now that the initial approver is created, the rest of the organisation must be verified. This can only be done by an SDE administrator, so will transition to the "Await pickup" state. From here onwards, the current request workflow continue as before.

## Registration Form

An anonymous user will be able to fill in a registration form in the Mauro Data Explorer, providing the following details:

- **Organisation name** - including the legal name
- **Organisation details** - website, country of origin, SMB status, etc
- **Department name** - to name the first department. If not provided, then assume the default name is the same as the organisation name.
- **Contact details** - personal details of the first user/organisation member:
    - **Name of user** - first and last
    - **Email address** - their organisation email address
    - **Telephone**
    - **Position/job title** - for records
- **Other contact details** - e.g. project proposals

Once submitted, the request will not currently be attached to a user ID - because the user does not exist yet. When the user record is available though, it will be assigned to the request for tracking purposes.

## SDE Administrator Actions

Once the registration request is picked up by an SDE administrator, the following actions must happen before an organisation is created and the first user/member can at a minimum sign-in to the Mauro Data Explorer and track their request:

1. **Create Mauro/SDE user accounts** - the SDE administrators must perform some initial checks to make sure a user account should be created e.g. the request is not spam or a bot. Once verified, some processes can happen to setup accounts. Note that these are no longer automatic, they must be triggered manually through the SDE Admin UI.
    1. **Mauro** - create a user account with the same email address as the contact details. Also assign the user to the correct user groups.
    2. **SDE** - create a user account with the same email address as the contact details.
2. **Create user in Azure** - next, a user account should be added/created in the Azure OUH tenant. This is a step performed by the OUH system administrators (currently not automated).
3. **Notification** - the user must be notified (somehow) that they can sign-in to the Mauro Data Explorer.
4. **Verification** - the SDE administrators/team must verify the suggested organisation and carry out "due diligence". These are not necessarily steps that can be automated in the SDE backend.
2. **Create organisation** - once verified, an automated step can happen to create a new organisation via the existing `CreateOrganisationCommand`. The first department will also automatically be created.

**Note:** although the user may sign-in finally and track their request, they must still be prevented from performing any actions until they are added to an Organisation/Department.

## Add to Organisation/Department

Whilst the research user (approver) is able to access the Mauro Data Explorer, the user registration request can continue to be fulfilled by the SDE administrator. Once verified and approved, the SDE administrator will be able to run the following commands:

1. Add user to the new organisation
2. Add user to the new department within the organisation

Only once these are done can the organisation registration request be considered "Complete". Now the researcher is an organisation approver and is able to help verify new users who submit _user registration requests_.

**Comment by pjmonks** [comment]
I will close this issue as it is now superseded by #445.
